### DISCUSSION TITLE
Boolean operations between a Mesh and junction hang

### DISCUSSION THREAD

**Discussion by jlvila** [jimy-byerley/pymadcad#31] [Q&A] [answered]
Dear,

I need to construct objects using boolean functions (`union()`, `difference()` ...). All work well between meshes built upon `brick()`, `tube()`, `uvsphere()`.

Now, when I try to union or difference between a brick() and a junction, I raise the exception **_TriangulationError("no more feasible triangles (algorithm failure or bad input outline)", [outline.indices[i] for i in hole])_**

Here is the image before union or difference:
![boolean](https://user-images.githubusercontent.com/8035931/158624637-0cdda41a-4728-4eaa-9092-6ac9cf3a4120.png)

here is my piece of code to generate junction:

```python
    def addCone(self, x, y, w, h, color: QColor):
        print(f'Add cone: ', x, y, w, h)
        _radius = w/2.0
        _length = h
        interfaces = [
            flatsurface(wire(Circle((vec3(0), vec3(0, 1, 0)), _radius))),
            flatsurface(wire(Circle((vec3(0, _length, 0), vec3(0, 1, 0)), 0.001)))
        ]
        _cone = junction(
            interfaces[0],
            interfaces[1],
            tangents='straight'
        )
        _coneTranslated = _cone.transform(
            vec3(x+_radius, y, 0))
        _coneTranslated.option(color=vec3(color.red()/255.0,
                                          color.green()/255.0, color.blue()/255.0))
```

Thanks for your works and help ;-)

**Comment by jimy-byerley** [comment, marked as answer]
Hello @jlvila 

I could not reproduce a `TriangulationError` with this piece of code, however I got wrong meshing, which can be due to the same cause. I guess it's only because I had not the same exact brick in the boolean operation. If fixed it by taking a closer look at the winding of the junction interfaces: your contours are both going the same way, which is not correct for a cone.
The interfaces of a junction will be used as surface contours in the final mesh, so each side must have opposed windings. If you flip one of your circles, then you should get a proper cone and solve `TriangulationError`

here is my working piece of code
```python
_brick = brick(center=1.5*X+2.5*Y, width=vec3(2))

x = 1
y = 1
w = 1
h = 1
color = vec3(1)

_radius = w/2.0
_length = h
interfaces = [
	flatsurface(wire(Circle((vec3(0), vec3(0, 1, 0)), _radius))),
	flatsurface(wire(Circle((vec3(0, _length, 0), vec3(0, -1, 0)), 0.001)))
]
_cone = junction(
	interfaces[0],
	interfaces[1],
	tangents='straight'
)
_coneTranslated = _cone.transform(vec3(x+_radius, y, 0))
_coneTranslated.option(color=vec3(color.r, color.g, color.b))
_coneTranslated.mergeclose()

result = difference(_brick, _coneTranslated)
```
![image](https://user-images.githubusercontent.com/7940753/159134331-8103cf97-701a-4f94-97e9-65bec848ae2f.png)

The only change is `Circle(... vec3(0,1,0), ...)` that becomes `Circle(..., vec3(0,-1,0), ...)`

By zooming on the cone angle, you will see the problem you had:

- same windings
   ![image](https://user-images.githubusercontent.com/7940753/159134455-432af076-9426-411a-b2f6-262e8e572d57.png)
- opposed windings
   ![image](https://user-images.githubusercontent.com/7940753/159134479-91890e2d-23f0-4ec5-ab3b-524d14d9a5db.png)

The same winding was creating some kind of paradoxical surface that went through itself, hence having no clear in- and out-side, and the boolean operations are not pleased when in and out are not consistent

**Reply by jlvila** [reply]
I just change the axis orientation, I've less exceptions but still have some

The main difference I have is the size of the objects, if I increase the radius of the cone end part  (0.1 in place of 0.001), it works. 

In your code change 0.001 to 0.00001 and you will have the exception
